src/test/clojure/clara/tools/test_tracing.cljc

```diff
@@ -0,0 +1,147 @@
+                       [cljs.test :refer [is deftest run-tests testing use-fixtures]])))
+
+(def-rules-test test-tracing-toggle
+  {:rules [cold-rule [[[Temperature (< temperature 20)]]
+                      (println "It's cold!")]]
+
+   :sessions [session [cold-rule] {}]}
+
+  (is (= false (t/is-tracing? session)))
+
+  (is (= true (t/is-tracing? (-> session
+                                 (t/with-tracing)))))
+
+  (is (= false (t/is-tracing? (-> session
+                                  (t/with-tracing)
+                                  (t/without-tracing))))))
+
+(def rule-output (atom nil))
+(def-rules-test test-simple-trace
+  {:rules [cold-rule [[[Temperature (< temperature 20)]]
+                      (reset! rule-output ?__token__)]]
+   :sessions [empty-session [cold-rule] {}]}
+
+  (let [session (-> empty-session
+                    (t/with-tracing)
+                    (insert (->Temperature 10 "MCI"))
+                    (fire-rules))]
+
+    ;; Ensure expected events occur in order.
+    (is (= [:add-facts :alpha-activate :right-activate :left-activate :add-activations]
+           (map :type (t/get-trace session))))))
+
+(def-rules-test test-rhs-retraction-trace
+  {:rules [cold-rule [[[Temperature (< temperature 20)]]
+                      (retract! (->Hot :too-hot))]]
+   :queries [hot-query [[]
+                        [[?hot <- Hot]]]]
+   :sessions [empty-session [cold-rule hot-query] {}]}
+
+  (let [session (-> empty-session
+                    (t/with-tracing)
+                    (insert (->Hot :too-hot))
+                    (fire-rules)
+                    (insert (->Temperature 10 "MCI"))
+                    (fire-rules))]
+    (is (= (map :type (t/get-trace session))
+           [:add-facts :alpha-activate :right-activate :left-activate
+            :add-facts :alpha-activate :right-activate :left-activate
+            :add-activations :retract-facts :alpha-retract :right-retract :left-retract])
+        "Validate that a retract! call in the RHS side of a rule appears in the trace
+         before the :right-retract")))
+        
+(def-rules-test test-accumulate-trace
+  {:queries [coldest-query [[]
+                            [[?t <- (acc/min :temperature :returns-fact true) from [Temperature]]]]]
+   :sessions [empty-session [coldest-query] {}]}
+
+  (let [session (-> empty-session
+                    (t/with-tracing)
+                    (insert (->Temperature 15 "MCI"))
+                    (insert (->Temperature 10 "MCI"))
+                    (insert (->Temperature 80 "MCI"))
+                    fire-rules)]
+
+    (is (= [:add-facts :alpha-activate :right-activate :accum-reduced
+            :left-activate :add-facts :alpha-activate :right-activate
+            :accum-reduced :left-retract :left-activate :add-facts
+            :alpha-activate :right-activate :accum-reduced]
+
+           (map :type (t/get-trace session))))))
```

Comment:
No, I think it just hasn't come up much.  It would perhaps be ideal to have some sort of editor-independent automatic formatter e.g. a Leiningen formatting plugin, but I haven't investigated that in any meaningful way.